### PR TITLE
Category filter

### PR DESCRIPTION
**Pull request by xaviqv**
This PR is build upon another pending PR "New agent and owner menu" https://github.com/thekordy/ticketit/pull/322

Added category filter on the same header menu. About this update:
- Listed agents depending on filtered category.
- Ticket list titles specify category, agent and owner (if any of these is chosen).
- Category column disappears when category filter is in use

### CODE REVIEW

**Review by munkhsaikhans — APPROVED** [review]

File: src/Controllers/TicketsController.php [hunk 1]
```diff
@@ -32,24 +32,23 @@ public function data(Datatables $datatables, $complete = false)
...
-            } else {
-                $collection = Ticket::active();
-            }
-        } elseif ($user->isAgent()) {
-            if ($complete) {
-                $collection = Ticket::complete()->agentUserTickets($user->id);
-            } else {
-                $collection = Ticket::active()->agentUserTickets($user->id);
-            }
+        $collection = Ticket::listComplete($complete);
```
Comment (balping) [inline src/Controllers/TicketsController.php:35]: `$collection` doesn't really describe the purpose of this variable.

File: src/Models/Agent.php [hunk 2]
```diff
@@ -270,4 +270,47 @@ public function userOpenTickets()
...
+     */
+    public function scopeVisibleForAgent($query, $id)
+    {
+        // Depends on agent_restrict
+        if (Setting::grab('agent_restrict') == 0) {
+            return $query->whereHas('categories', function ($q1) use ($id) {
+                $q1->whereHas('agents', function ($q2) use ($id) {
+                    $q2->where('id', $id);
+                });
+            })->orderBy('name', 'ASC');
```
Comment (balping) [inline src/Models/Agent.php:311]: Code duplication

File: src/Models/Agent.php [hunk 3]
```diff
@@ -270,4 +270,47 @@ public function userOpenTickets()
...
+     * @return mixed
+     */
+    public function scopeVisible($query)
+    {
+        $query = $query->agents();
+
+        if (auth()->user()->ticketit_admin) {
+            return $query->orderBy('name', 'ASC');
+        } elseif (auth()->user()->ticketit_agent) {
+            return $query->VisibleForAgent(auth()->user()->id);
```
Comment (balping) [inline src/Models/Agent.php:289]: Let's stick to the convention of having the first letter of scopes lower case. Applies elsewhere too

File: src/Models/Ticket.php [hunk 4]
```diff
@@ -189,6 +203,66 @@ public function scopeAgentUserTickets($query, $id)
...
+
+    /**
+     * Get all tickets in specified category.
+     *
+     * @param $query
+     * @param $id
+     *
+     * @return mixed
+     */
+    public function scopeInCategory($query, $id = null)
```
Comment (balping) [inline src/Models/Ticket.php:256]: Should we move these whole bunch of scopes to a trait maybe?

File: src/Views/shared/assets.blade.php [hunk 5]
```diff
@@ -1,4 +1,7 @@
 {{-- Load the css file to the header --}}
+<link rel="StyleSheet" href="{{asset('vendor/ticketit/css/select2.min.css')}}">
```
Comment (balping) [inline src/Views/shared/assets.blade.php:2]: I'd prefer pulling this jquery plugin via cdn, as we do for codemirror, summernote and fontawesome. No need to keep a copy of this library in this repo + it's easier to upgrade.

**Review by balping — CHANGES_REQUESTED** [review]
I haven't really reviewed the blade files, but the design is ok, I guess

File: src/Controllers/TicketsController.php (hunk 1 above)
Comment (balping) [reply src/Controllers/TicketsController.php:35]: I can see it was like that before, but maybe it's time to change it.

**Comment by thekordy** [comment]
A lot of issues, let's move on.

**Pull request closed without merging**